src/nbl/system/ISystem.cpp

```diff
@@ -7,4 +7,46 @@ namespace nbl::system
     {
         addArchiveLoader(core::make_smart_refctd_ptr<CArchiveLoaderZip>());
     }
+    core::smart_refctd_ptr<IFile> ISystem::getFileFromArchive(const system::path& _path)
+    {
+        system::path path = _path.parent_path().string();
+
+        
+        bool isPathAlias = !std::filesystem::exists(path);
+        while (!path.empty()) // going up the directory tree
+        {
+            system::path realPath = path;
+            if (isPathAlias)
+            {
+                auto a = m_cachedPathAliases.findRange(path);
+                if (a.empty())
+                {
+                    path = path.parent_path();
+                    continue;
+                }
+                realPath = a.begin()->second;
+            }
+            auto archives = m_cachedArchiveFiles.findRange(realPath);
+            for (auto& archive : archives)
+            {
+                auto relative = std::filesystem::relative(_path, path);
+                auto files = archive.second->getArchivedFiles();
+                auto requiredFile = std::find_if(files.begin(), files.end(), [&relative](const IFileArchive::SFileListEntry& entry) { return entry.fullName == relative; });
+                if (requiredFile != files.end()) return archive.second->readFile({ relative, "" }); //TODO password
```

Comment:
the file you return should be cached internally by the archive.
(so that you dont spam a new file every time someone gets it)

Here's how to do it:
1. Change https://github.com/Devsh-Graphics-Programming/Nabla/blob/bca63c5737d56479d312a4fea7853ad5a94ef006/include/nbl/core/IReferenceCounted.h#L65 to return the value of `ReferenceCounter++`
2. Reserve an array of `std::atomic_flag` elements inside the File Archive sufficient to have one entry per entry in the file list and clear the array to `false`
3. Derive an `IArchiveFile` from `IFile`, hold a pointer to an element of the array in (2) and have the constructor set the element to `true` and destructor to `false`
4. Reserve an array of `sizeof(IFile)` elements of memory inside the File Archive sufficient to have one entry per entry in the file list and clear the memory with zeros
5. Protect the array from (4) with a mutex

When you get a request for a file:
1. Translate relative path to file Index
2. look at the correct `IFile` array index (then cast the memory to `IFile`)
3. lock the mutex (very important it happens here)
4. Call `grab()` on the IArchiveFile (doesn't matter if hasn't been constructed yet)
5. If `grab()` returned a value greater than 0, unlock, construct a smartpointer to the `IArchiveFile` and return it (but also assert that the atomic_flag is true)

6. If `grab()` returned value equal to 0, then no one else was holding the file (means file not initialized, has died or will die)
7. C++20 wait on the  `atomic_flag` to transition to `false` (see the AsyncDispatchQueue for an example how to wait for such a transition robustly)
8. Placement-new an `IArchiveFile` over the memory (this will overwrite the refcount), give the atomic_flag address as a parameter
9. construct smart pointer with `core::dont_grab`,unlock the mutex and return the smart pointer